### PR TITLE
[AIRFLOW-3276] Cloud SQL: database insert / patch / delete

### PR DESCRIPTION
**Pull request by sprzedwojski**
Add CloudSqlInstanceDatabaseInsertOperator, CloudSqlInstanceDatabasePatchOperator and CloudSqlInstanceDatabaseDeleteOperator.

Each operator includes:

- core logic
- input params validation
- unit tests
- presence in the example DAG
- docstrings
- How-to and Integration documentation

### CODE REVIEW

File: airflow/contrib/example_dags/example_gcp_sql.py [hunk 1]
```diff
@@ -123,6 +139,45 @@
...
+    # [START howto_operator_cloudsql_db_insert]
+    sql_db_insert_task = CloudSqlInstanceDatabaseInsertOperator(
+        project_id=PROJECT_ID,
+        body=db_insert_body,
+        instance=INSTANCE_NAME,
+        task_id='sql_db_insert_task'
+    )
+    # [END howto_operator_cloudsql_db_insert]
+    # A second, identical database insert to verify idempotence
+    sql_db_insert_task_2 = CloudSqlInstanceDatabaseInsertOperator(
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/example_dags/example_gcp_sql.py]: You're quite right, the doubled operators' invocations in the example DAG may be confusing for users.
Moreover, I realised that idempotence in our cases is about not performing an operation in GCP under certain conditions (insert - if sth already exists, delete - if sth doesn't exist), so it can in fact be tested with unit tests I guess and that should be enough.

Therefore I'll remove the doubled operators from the DAG and I'll add any missing unit tests to cover this.

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 2]
```diff
@@ -144,6 +144,97 @@ def delete_instance(self, project_id, instance):
...
         return self._wait_for_operation_to_complete(project_id, operation_name)
 
+    def get_database(self, project_id, instance, database):
+        """
+        Retrieves a resource containing information about a database inside a Cloud SQL
+        instance.
+
+        :param project_id: Project ID of the project that contains the instance.
+        :type project_id: str
+        :param instance: Database instance ID. This does not include the project ID.
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: This is copied 1:1 from the docs: https://developers.google.com/resources/api-libraries/documentation/sqladmin/v1beta4/python/latest/sqladmin_v1beta4.databases.html#delete

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 3]
```diff
@@ -144,6 +144,97 @@ def delete_instance(self, project_id, instance):
...
+        """
+        return self.get_conn().databases().get(
+            project=project_id,
+            instance=instance,
+            database=database
+        ).execute(num_retries=NUM_RETRIES)
+
+    def insert_database(self, project, instance, body):
+        """
+        Inserts a resource containing information about a database inside a Cloud SQL
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: Yeah, your suggestion is probably better, so I'll change it.

However it was again copied directly from the docs https://developers.google.com/resources/api-libraries/documentation/sqladmin/v1beta4/python/latest/sqladmin_v1beta4.databases.html#insert

File: airflow/contrib/example_dags/example_gcp_sql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/example_dags/example_gcp_sql.py]: Done, removed the doubled operators.

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 4]
```diff
@@ -144,6 +144,97 @@ def delete_instance(self, project_id, instance):
...
+        Retrieves a resource containing information about a database inside a Cloud SQL
+        instance.
+
+        :param project_id: Project ID of the project that contains the instance.
+        :type project_id: str
+        :param instance: Database instance ID. This does not include the project ID.
+        :type instance: str
+        :param database: Name of the database in the instance.
+        :type database: str
+        :return: A Cloud SQL instance resource.
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: Corrected

File: airflow/contrib/hooks/gcp_sql_hook.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: Done

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 5]
```diff
@@ -144,6 +144,97 @@ def delete_instance(self, project_id, instance):
...
+            project=project,
+            instance=instance,
+            body=body
+        ).execute(num_retries=NUM_RETRIES)
+        operation_name = response["name"]
+        return self._wait_for_operation_to_complete(project, operation_name)
+
+    def patch_database(self, project, instance, database, body):
+        """
+        Updates a resource containing information about a database inside a Cloud SQL
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: This is coming from the docs, but I changed it according to your earlier suggestion.

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 6]
```diff
@@ -286,3 +334,184 @@ def execute(self, context):
...
+    :param project_id: Project ID of the project that contains the instance.
+    :type project_id: str
+    :param instance: Database instance ID. This does not include the project ID.
+    :type instance: str
+    :param body: The request body, as described in
+        https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/databases/insert#request-body
+    :type body: dict
+    :param gcp_conn_id: The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param api_version: API version used (e.g. v1).
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py]: Oops, thanks :) Changed

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 7]
```diff
@@ -286,3 +334,184 @@ def execute(self, context):
...
+    :param instance: Database instance ID. This does not include the project ID.
+    :type instance: str
+    :param database: Name of the database to be updated in the instance.
+    :type database: str
+    :param body: The request body, as described in
+        https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/databases/patch#request-body
+    :type body: dict
+    :param gcp_conn_id: The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param api_version: API version used (e.g. v1).
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py]: Changed

File: docs/howto/operator.rst [hunk 8]
```diff
@@ -284,6 +284,148 @@ If the source code for your function is in Google Source Repository, make sure t
...
+.. literalinclude:: ../../airflow/contrib/example_dags/example_gcp_sql.py
+    :language: python
+    :start-after: [START howto_operator_cloudsql_db_insert_body]
+    :end-before: [END howto_operator_cloudsql_db_insert_body]
+
+Templating
+""""""""""
+
+.. literalinclude:: ../../airflow/contrib/operators/gcp_sql_operator.py
+  :language: python1
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply docs/howto/operator.rst]: Sorry, bad mistake. Corrected

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 9]
```diff
@@ -286,3 +334,184 @@ def execute(self, context):
...
+                           " database already exists in the instance {}."
+                           .format(self.instance))
+            return False
+        if self._check_if_db_exists(database):
+            self.log.info("Cloud SQL instance with ID {} already contains database"
+                          " '{}'. Aborting database insert."
+                          .format(self.instance, database))
+            return True
+        else:
+            return self._hook.insert_database(self.project_id, self.instance, self.body)
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py]: I tried running some semi-manual concurrency tests, using our example DAG and the gcloud CLI, and generally as soon as the Cloud SQL instance receives an `insert database` command, the `gcloud database describe` starts returning the database resource information.

So in your example I guess as soon as TI1 issues the `insert` call, any `_check_if_db_exists ` invocations should return `True` (even if the database will really be accessible a few seconds later, as this is what it takes to create the database).

**TL;DR**
If a database is being created, the API reports it as an existing database.

Therefore I don't think there are really any concurrency-related issues in play here.
Feel free to correct me if I'm wrong :)

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 10]
```diff
@@ -286,3 +334,184 @@ def execute(self, context):
...
+            GcpBodyFieldValidator(CLOUD_SQL_DATABASE_PATCH_VALIDATION,
+                                  api_version=self.api_version).validate(self.body)
+
+    def execute(self, context):
+        self._validate_body_fields()
+        if not self._check_if_db_exists(self.database):
+            self.log.warning("Cloud SQL instance with ID {} does not contain "
+                             "database '{}'. Aborting database patch."
+                             .format(self.instance, self.database))
+            return False
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py]: You're right. Raising an exception here is also consistent with the instance patch operator.
Changed.

File: tests/contrib/operators/test_gcp_sql_operator.py [hunk 11]
```diff
@@ -312,3 +326,112 @@ def test_instance_delete_should_abort_and_succeed_if_not_exists(
...
+        _check_if_db_exists.return_value = False
+        op = CloudSqlInstanceDatabasePatchOperator(
+            project_id=PROJECT_ID,
+            instance=INSTANCE_NAME,
+            database="db1",
+            body=DATABASE_INSERT_BODY,
+            task_id="id"
+        )
+        result = op.execute(None)
+        self.assertFalse(result)
```
[the remark this answers is not in the document]
Comment (sprzedwojski) [reply tests/contrib/operators/test_gcp_sql_operator.py]: Done

**Comment by sprzedwojski** [comment]
@TrevorEdwards I've addressed all of your comments. As usual, a very good CR, thanks ;)
In most cases I've altered the code according to your suggestions, however in a few cases I just left a comment here.
Would be great if you could take a look when you have a moment :)

File: airflow/contrib/operators/gcp_sql_operator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TrevorEdwards) [reply airflow/contrib/operators/gcp_sql_operator.py]: SGTM

File: docs/howto/operator.rst [hunk 12]
```diff
@@ -284,6 +284,148 @@ If the source code for your function is in Google Source Repository, make sure t
...
+
+Creates a new database inside a Cloud SQL instance.
+
+For parameter definition take a look at
+:class:`~airflow.contrib.operators.gcp_sql_operator.CloudSqlInstanceDatabaseCreateOperator`.
+
+Arguments
+"""""""""
+
+Some arguments in the example DAG are taken from Airflow variables:
```
Comment (TrevorEdwards) [inline docs/howto/operator.rst]: environment variables

File: docs/howto/operator.rst [hunk 13]
```diff
@@ -284,6 +284,148 @@ If the source code for your function is in Google Source Repository, make sure t
...
+
+Deletes a database from a Cloud SQL instance.
+
+For parameter definition take a look at
+:class:`~airflow.contrib.operators.gcp_sql_operator.CloudSqlInstanceDatabaseDeleteOperator`.
+
+Arguments
+"""""""""
+
+Some arguments in the example DAG are taken from Airflow variables:
```
Comment (TrevorEdwards) [inline docs/howto/operator.rst]: environment

File: docs/howto/operator.rst [hunk 14]
```diff
@@ -284,6 +284,148 @@ If the source code for your function is in Google Source Repository, make sure t
...
+using patch semantics.
+See: https://cloud.google.com/sql/docs/mysql/admin-api/how-tos/performance#patch
+
+For parameter definition take a look at
+:class:`~airflow.contrib.operators.gcp_sql_operator.CloudSqlInstanceDatabasePatchOperator`.
+
+Arguments
+"""""""""
+
+Some arguments in the example DAG are taken from Airflow variables:
```
Comment (TrevorEdwards) [inline docs/howto/operator.rst]: environment

File: docs/howto/operator.rst (hunk 12 above)
Comment (sprzedwojski) [reply docs/howto/operator.rst]: Thanks! Corrected this for all 6 Cloud SQL operators in the `operator.rst`. Also I noticed that the example DAG's docstring was out of date, so I updated it too.

**Comment by sprzedwojski** [comment]
I'm closing this PR and will open a new one to the `apache/incubator-airflow` repo.
